### ISSUE TITLE
Go to definition for the second time fails

### ISSUE THREAD

**Issue by kamilez** [prabirshrestha/vim-lsp#97]
Hello, I am using vim 8.0 with vim-lsp and dart_language_server and found very disturbing issue. When the file is loaded and for the first time go to definition of the word under the cursor, it works fine. But then whenever I go back with (ctrl-o) and once again try to go to the definition of the same word it hangs on 'Retrieving definition ...' on other words, works fine but still only once. Because I don't know vimscript very well all I could find is that the second time it hangs on getbufline(bufnr(l:path), l:line)[0] in ui/vim/utils.vim:16. Another issue that might be related with this is when I had set

autocmd FileType dart setlocal shiftwidth=2 tabstop=2 expandtab

in my .vimrc file every time on go to definition and back, the shiftwidth was set to 8 (default value?), no idea why and had to change the line above from setlocal to just set to make it fixed. Is it some kind of buffer issue? How to fix it. Thank you in advance.

**Comment by natebosch** [comment]
What version of dart_language_server are you using? I've had some bugs where certain calls can lock it up and it stops working for a given file. I've fixed that bug but only in versions available for the 2.0.0-dev sdks.

The shiftwidth issue would not be related. You might have some other config setting it on something other autocmd or something - does it happen if you jump to another file with `:edit` then back? I think you can also see where it was set from last with `:set verbose shiftwidth?`

**Comment by kamilez** [comment]
dart sdk it is 2.0.0-dev.7.0
dart_language_server 0.1.5

With this shiftwidth issue you were right. I removed all plugins end setup leaving only those for dart and now it does not change.

**Comment by natebosch** [comment]
The latest for dart_language_server is 0.1.7 - please do another `pub global activate` to get that version. It should resolve the issue with requests getting blocked and never responding.

**Comment by kamilez** [comment]
Well, I updated the dart_language_server to 0.1.7 and still this didn't help. Replaced for a while vim-lsp with vim-lsc and the issue seems to not appear there, but would like to stay with vim-lsp.

**Comment by natebosch** [comment]
Ah cool, in that case it's either a bug in `dart_language_server` triggered by some behavior of vim-lsp that I don't have in vim-lsc, or a bug in vim-lsp triggered by some behavior of the server :)

@prabirshrestha might be able to help with specifics around diagnosing the state of vim-lsp.

If you can record some communication logs with the server when it gets stuck (either set the trace level to 'message' or 'verbose' during the initialization call, or wrap the server with something that logs with `tee`) we can dig from there. I'm happy to fix any protocol bugs in the server.

**Comment by kamilez** [comment]
[vim_server_log.txt](https://github.com/prabirshrestha/vim-lsp/files/1649340/vim_server_log.txt)

Here are logs from tee. Only the last line is from the point when go to definition hangs.

**Comment by natebosch** [comment]
The last message coming back does look like it's a valid result for a `textDocument/definition` request, and both id 2 and id 3 calls match this.
`{"jsonrpc":"2.0","result":{"range":{"end":{"character":12,"line":128},"start":{"character":6,"line":128}},"uri":"file:///home/kamil/Flutter/packages/flutter/lib/src/material/app_bar.dart"},"id":3}`

So I think the server did see and respond correctly to both calls. I suspect the issue is in the client.

There are some very long messages coming form the server in this log. Perhaps being hit by the same issue as https://github.com/natebosch/vim-lsc/issues/67, but probably not since the response to the definition request is coming through.

At the version you're using the server size should be chunking writes. If vim-lsp isn't also chunking writes then things could still get stuck. The [file you jumped to](https://github.com/flutter/flutter/blob/355ebef5fede5db56e9f7f59e24279635af28a71/packages/flutter/lib/src/material/app_bar.dart#L129) is 34.7k so the `didOpen` call would be at least that big. Some of the messages coming from the server were >50k. These messages might be bigger than the pipe buffer (could be as small as 4k on older linux kernels, I _think_ defaults to 16k on mac?), but I don't see the telltale cutoff message and the behavior doesn't quite match what I'd expect if this was the problem.

**Comment by kamilez** [comment]
Here is a log from lsp

nie, 21 sty 2018, 15:22:31:[{"response":{"data":{"__data__":"vim-lsp","server_name":"dart_language_server_tee"},"message":"server already started"}}]
nie, 21 sty 2018, 15:22:31:[{"response":{"data":{"__data__":"vim-lsp","init_result":{"id":1,"jsonrpc":"2.0","result":{"capabilities":{"executeCommandProvider":null,"hoverProvider":true,"workspaceSymbolProvider":false,"referencesProvider":true,"signatureHelpProvider":null,"codeActionProvider":true,"textDocumentSync":{"save":false,"willSaveWaitUntil":false,"willSave":false,"change":2,"openClose":true},"codeLensProvider":false,"documentHighlightsProvider":true,"documentLinkProvider":null,"documentOnTypeFormattingProvider":false,"definitionProvider":true,"documentRangeFormattingProvider":false,"documentFormattingProvider":false,"documentSymbolProvider":false,"renameProvider":true,"completionProvider":{"resolveProvider":false,"triggerCharacters":["."]}}}},"server_name":"dart_language_server_tee"},"message":"lsp server already initialized"}}]
nie, 21 sty 2018, 15:22:31:[{"response":{"data":{"path":"file:///home/kamil/workspace/district/lib/views/notifications_view.dart","__data__":"vim-lsp","server_name":"dart_language_server_tee"},"message":"already opened"}}]
nie, 21 sty 2018, 15:22:31:["--->",1,"dart_language_server_tee",{"method":"textDocument/didChange","params":{"contentChanges":[{"text":"import 'package:flutter/material.dart';\nimport 'package:intl/intl.dart';\nimport './flexible_gradient.dart';\nimport './tab_view.dart';\nimport './utils.dart';\n\nclass NotificationsView extends StatefulWidget implements TabViewItem {\n  @override\n  IconData get icon => Icons.notifications;\n\n  @override\n  String get name => \"Notifications\";\n\n  @override\n  _NotificationsViewState createState() => new _NotificationsViewState();\n}\n\nclass _NotificationsViewState extends State<NotificationsView> {\n  @override\n  Widget build(BuildContext context) {\n    final listItems = new List<Widget>();\n\n    for (var i = 0; i < 20; i++) {\n      listItems.add(new IconTextListItemView(\n        \"Someone left comment in your post\" + i.toString(),\n        \"images/user.png\",\n        subText: new DateFormat(\"d MMM y\").format(new DateTime.now()),\n      ));\n    }\n\n    return new Scaffold(\n      appBar: new AppBar(\n        elevation: 1.0,\n        flexibleSpace: new FlexibleGradient(),\n        title: const Text(\"Notifications\"),\n      ),\n      body: new Container(\n        color: Colors.grey.shade300,\n        child: new ListView(\n          children: listItems.map((Widget item) {\n            return item;\n          }).toList()\n        ),\n      ),\n    );\n  }\n}"}],"textDocument":{"uri":"file:///home/kamil/workspace/district/lib/views/notifications_view.dart","version":3}}}]
nie, 21 sty 2018, 15:22:31:[{"response":{"data":{"path":"file:///home/kamil/workspace/district/lib/views/notifications_view.dart","__data__":"vim-lsp","server_name":"dart_language_server_tee"},"message":"textDocument/didChange sent"}}]
nie, 21 sty 2018, 15:22:31:["--->",1,"dart_language_server_tee",{"method":"textDocument/definition","on_notification":"---funcref---","params":{"textDocument":{"uri":"file:///home/kamil/workspace/district/lib/views/notifications_view.dart"},"position":{"character":18,"line":31}}}]
nie, 21 sty 2018, 15:22:31:["<---",1,"dart_language_server_tee",{"response":{"id":3,"jsonrpc":"2.0","result":{"uri":"file:///home/kamil/Flutter/packages/flutter/lib/src/material/app_bar.dart","range":{"end":{"character":12,"line":128},"start":{"character":6,"line":128}}}},"request":{"method":"textDocument/definition","jsonrpc":"2.0","id":3,"params":{"textDocument":{"uri":"file:///home/kamil/workspace/district/lib/views/notifications_view.dart"},"position":{"character":18,"line":31}}}}]
nie, 21 sty 2018, 15:22:31:["s:on_stdout client request on_notification() error","Vim(let):E684: list index out of range: 0"]

**Comment by prabirshrestha** [comment]
@kamilez I have seen this too with typescript language sever so it is defentily a bug in vim-lsp. can you try reverting a2483c71d168bae91ac1426984613b50add7a81f and check if it works?

**Comment by kamilez** [comment]
Yes. Reverting solves this issue, but is it a final solution?

**Comment by prabirshrestha** [comment]
@kamilez It is ok to revert for now since the PR that introduced the bug is an enhancement. https://github.com/prabirshrestha/vim-lsp/pull/93.
